### PR TITLE
Improve Ersatz.Relation

### LINKED ISSUE #71: improve Ersatz.Relation (docs, features) - or break out?
I have a student working on Ersatz.Relation, and some doc fixes incoming (this week). 

There will also be additions (properties of relations that are used in abstract rewriting systems) - perhaps this is best done in a separate package (but it would be quite small).

Should Ersatz.Relation be broken out of ersatz as well? (It was not there to begin with.) This would reduce cognitive load for users and maintainers of  the core library - and reduce churn of version numbers! (by PVP, have to update on API changes)

### PR DESCRIPTION
**Pull request by aswiridoff**
resolve #71 (?)

### CODE REVIEW

**Comment by RyanGlScott** [comment]
Thanks for the patch, @aswiridoff! Some of the `hlint` suggestions (in the 9.0.2 CI job) are rather silly, so I've disabled a subset of them. Can you resolve the remaining `hlint` suggestsions?

Let me know when you'd like a full review.

**Comment by aswiridoff** [comment]
@RyanGlScott Thanks! I've resolved the remaining suggestions. I think it should be ready to merge now.

File: src/Ersatz/Relation/ARS.hs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+-- | Tests if a relation \( R \subseteq A \times A \) is terminating, i.e., 
+-- there is no infinite sequence \( x_1, x_2, ... \) with \( x_i \in A \)
+-- such that \( (x_i, x_{i+1}) \in R \) holds.
+--
+-- Formula size: linear in \( |A|^3 \)
+terminating :: Ix a => Relation a a -> Bit
+terminating r = irreflexive $ transitive_closure r
+
+-- | Monadic version of @'terminating'@.
```
[the remark this answers is not in the document]
Comment (aswiridoff) [reply src/Ersatz/Relation/ARS.hs]: Thanks for the hint!

File: src/Ersatz/Relation/Data.hs [hunk 2]
```diff
@@ -79,38 +80,53 @@ symmetric_relation bnd = do
...
 build bnd pairs = Relation $ A.array bnd pairs
 
 -- | Constructs a relation \(R \subseteq A \times B \) from a function.
 buildFrom :: (Ix a, Ix b)
-          => (a -> b -> Bit) -- ^ A function with the specified signature, that assigns a 'Bit'-value 
-                             -- to each element \( (x,y) \in A \times B \).
-          -> ((a,b),(a,b))
+          => ((a,b),(a,b))
+          -> ((a,b) -> Bit) -- ^ A function that assigns a @Bit@-value 
+                            -- to each element \( (x,y) \in A \times B \).
```
[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Ersatz/Relation/Data.hs]: I don't have a strong opinion on what the order should be, just that we announce the new type somewhere in the `CHANGELOG` :)

[the remark this answers is not in the document]
Comment (aswiridoff) [reply src/Ersatz/Relation/Data.hs]: Yes, I'm sorry it took so long! Do you think it's fine like this? 721f3082a3a9ff0260a4e4b4e88d2765caa3f6d4
I'm not quite sure how the version numbering works.

[the remark this answers is not in the document]
Comment (RyanGlScott) [reply src/Ersatz/Relation/Data.hs]: That looks great, thank you! If you'd like, you could also add something like this to the top of the file:

```
next [????.??.??]
-----------------
```

(I can fill in the rest when it's time for a new `ersatz` release.)

Also, can you resolve the merge conflict with the `master` branch?

**Review by RyanGlScott — APPROVED** [review]
Thanks again!

**Comment by aswiridoff** [comment]
Thanks a lot for your feedback, @RyanGlScott! :)

**Pull request merged**